pyresample/gradient/__init__.py

```diff
@@ -22,36 +22,60 @@
 # with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 """Implementation of the gradient search algorithm as described by Trishchenko."""
+from __future__ import annotations
 
 import logging
+from functools import wraps
 
 import dask
 import dask.array as da
 import numpy as np
 import pyproj
 import xarray as xr
+from pyproj import CRS
 from shapely.geometry import Polygon
 
 from pyresample import CHUNK_SIZE
-from pyresample.geometry import get_geostationary_bounding_box
-from pyresample.gradient._gradient_search import one_step_gradient_search
+from pyresample.geometry import (
+    AreaDefinition,
+    SwathDefinition,
+    get_geostationary_bounding_box,
+)
+from pyresample.gradient._gradient_search import (
+    one_step_gradient_indices,
+    one_step_gradient_search,
+)
 from pyresample.resampler import BaseResampler
 
 logger = logging.getLogger(__name__)
 
 
+class GradientSearchResampler(BaseResampler):
+    """GradientSearch factory class."""
+
+    def __new__(cls, source_geo_def, target_geo_def):
+        """Create a GradientSearchResampler."""
+        if cls is GradientSearchResampler:
+            if isinstance(source_geo_def, AreaDefinition) and isinstance(target_geo_def, AreaDefinition):
+                return RBGradientSearchResampler(source_geo_def, target_geo_def)
+            elif isinstance(source_geo_def, SwathDefinition) and isinstance(target_geo_def, AreaDefinition):
+                return RBGradientSearchResampler(source_geo_def, target_geo_def)
+        else:
+            return super().__new__(cls)
```

Comment:
In addition to the discussion on slack of my displeasure with this implementation, doesn't this return `None` of the source target aren't area/area or swath/area?